File: trip-kit-booking-ui/src/main/java/com/skedgo/android/bookingclient/viewmodel/ExternalProviderAuthViewModel.java
package com.skedgo.android.bookingclient.viewmodel;

import android.content.Intent;
import android.databinding.ObservableBoolean;
import android.databinding.ObservableField;
import android.net.Uri;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.annotation.Nullable;
import android.support.annotation.VisibleForTesting;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import com.skedgo.android.bookingclient.OAuth2CallbackHandler;
import com.skedgo.android.tripkit.booking.BookingForm;

import javax.inject.Inject;

import rx.Observable;
import rx.Subscriber;
import rx.functions.Action1;
import rx.functions.Func1;
import rx.subjects.PublishSubject;
import skedgo.common.BaseViewModel;

import static com.skedgo.android.bookingclient.activity.ExternalProviderAuthActivity.KEY_BOOKING_FORM;

public class ExternalProviderAuthViewModel extends BaseViewModel {
  private final ObservableBoolean showWebView;
  private final PublishSubject<Intent> publishSubjectIntent = PublishSubject.create();
  private final ObservableField<String> url;

  @Nullable @VisibleForTesting BookingForm bookingForm;

  @Inject ExternalProviderAuthViewModel() {
    showWebView = new ObservableBoolean(false);
    url = new ObservableField<>();
  }

  public ObservableBoolean showWebView() {
    return showWebView;
  }

  public Observable<Intent> intentObservable() {
    return publishSubjectIntent.asObservable();
  }

  public ObservableField<String> url() {
    return url;
  }

  public void handleArgs(Bundle args) {
    BookingForm form = args.getParcelable(KEY_BOOKING_FORM);
    handleArgs(form);
  }

  private void handleArgs(BookingForm form) {
    this.bookingForm = form;

    if (bookingForm != null) {
      if (bookingForm.isOAuthForm() && bookingForm.getOAuthLink() != null) {
        url.set(bookingForm.getOAuthLink().toString());
      } else if (bookingForm.externalAction() != null) {
        url.set(bookingForm.externalAction());
      }
    }
  }

  public WebViewClient webViewClient(final OAuth2CallbackHandler oAuth2CallbackHandler) {
    return new WebViewClient() {
      @Override
      public boolean shouldOverrideUrlLoading(WebView webView, String url) {
        return handleCallback(url, oAuth2CallbackHandler);
      }

      @Override
      public void onPageFinished(WebView view, final String url) {
        if (!url.startsWith("tripgo://")) {
          showWebView.set(true);
        }
      }
    };
  }

  @VisibleForTesting
  boolean handleCallback(String webUrl, OAuth2CallbackHandler oAuth2CallbackHandler) {
    if (webUrl.startsWith("tripgo://")) {
      if (bookingForm != null) {
        if (webUrl.startsWith("tripgo://oauth-callback")) {
          String callback = webUrl.substring(0, webUrl.indexOf("?"));

          oAuth2CallbackHandler.handleOAuthURL(bookingForm, Uri.parse(webUrl), callback)
              .takeUntil(destroyed())
              .flatMap(new Func1<BookingForm, Observable<Intent>>() {
                @Override public Observable<Intent> call(BookingForm bookingForm) {
                  return handledForm(bookingForm);
                }
              })
              .subscribe(new Action1<Intent>() {
                @Override public void call(Intent intent) {
                  publishSubjectIntent.onNext(intent);
                }
              }, new Action1<Throwable>() {
                @Override public void call(Throwable throwable) {
                  publishSubjectIntent.onError(throwable);
                }
              });
        } else if (webUrl.startsWith("tripgo://booking_retry")) {
          oAuth2CallbackHandler.handleRetryURL(bookingForm, Uri.parse(webUrl))
              .takeUntil(destroyed())
              .flatMap(new Func1<BookingForm, Observable<Intent>>() {
                @Override public Observable<Intent> call(BookingForm bookingForm) {
                  return handledForm(bookingForm);
                }
              })
              .subscribe(new Action1<Intent>() {
                @Override public void call(Intent intent) {
                  publishSubjectIntent.onNext(intent);
                }
              }, new Action1<Throwable>() {
                @Override public void call(Throwable throwable) {
                  publishSubjectIntent.onError(throwable);
                }
              });
        }
      }
      showWebView.set(false);
      return false;
    } else {
      showWebView.set(true);
      this.url.set(webUrl);
    }
    return true;
  }

  @VisibleForTesting
  Observable<Intent> handledForm(final BookingForm form) {

    return Observable.create(new Observable.OnSubscribe<Intent>() {
      @Override public void call(Subscriber<? super Intent> subscriber) {

        // if it's external action, keep loading the web view
        if (form != null && (form.isOAuthForm() || form.externalAction() != null)) {
          handleArgs(form);
        } else {
          final Intent data = new Intent();
          // Form can be null, indicating booking end (auth case)
          data.putExtra(KEY_BOOKING_FORM, (Parcelable) form);
          subscriber.onNext(data);
        }
        subscriber.onCompleted();
      }
    });
  }
}

File: trip-kit-booking-ui/src/main/java/com/skedgo/tripkit/booking/ui/viewmodel/ExternalProviderAuthViewModel.kt
package com.skedgo.tripkit.booking.ui.viewmodel

import android.content.Intent
import androidx.databinding.ObservableBoolean
import androidx.databinding.ObservableField
import android.net.Uri
import android.os.Bundle
import android.os.Parcelable
import androidx.annotation.VisibleForTesting
import android.webkit.WebView
import android.webkit.WebViewClient
import com.skedgo.tripkit.booking.BookingForm
import com.skedgo.tripkit.booking.ui.OAuth2CallbackHandler
import com.skedgo.tripkit.booking.ui.activity.KEY_FORM
import io.reactivex.Observable
import io.reactivex.subjects.PublishSubject
import javax.inject.Inject

class ExternalProviderAuthViewModel @Inject internal constructor() : DisposableViewModel() {

  private val publishSubjectIntent = PublishSubject.create<Intent>()
  val url: ObservableField<String> = ObservableField()
  val showWebView: ObservableBoolean = ObservableBoolean(false)
  val intentObservable: Observable<Intent> = publishSubjectIntent.hide()

  @VisibleForTesting var bookingForm: BookingForm? = null

  fun handleArgs(args: Bundle) {
    val form = args.getParcelable<BookingForm>(KEY_FORM)
    handleArgs(form)
  }

  fun webViewClient(oAuth2CallbackHandler: OAuth2CallbackHandler): WebViewClient {
    return object : WebViewClient() {
      override fun shouldOverrideUrlLoading(webView: WebView, url: String): Boolean {
        return handleCallback(url, oAuth2CallbackHandler)
      }

      override fun onPageFinished(view: WebView, url: String) {
        showWebView.set((!url.startsWith("tripgo://")))
      }
    }
  }

  @VisibleForTesting
  fun handleCallback(webUrl: String, oAuth2CallbackHandler: OAuth2CallbackHandler): Boolean {

    with(webUrl) {
      return when {
        startsWith("tripgo://") -> {
          bookingForm?.let {
            when {
              startsWith("tripgo://oauth-callback") -> {
                val callback = substring(0, indexOf("?"))
                oAuth2CallbackHandler.handleOAuthURL(bookingForm!!, Uri.parse(this), callback)
              }
              startsWith("tripgo://booking_retry") -> oAuth2CallbackHandler.handleRetryURL(bookingForm!!, Uri.parse(this))
              else -> null
            }
                ?.takeUntil(onDispose())
                ?.flatMap { bookingForm -> handledForm(bookingForm) }
                ?.subscribe({ intent -> publishSubjectIntent.onNext(intent) }) { throwable ->
                  publishSubjectIntent.onError(throwable)
                }
          }
          showWebView.set(false)
          false
        }
        else -> {
          showWebView.set(true)
          url.set(this)
          true
        }
      }
    }

  }

  @VisibleForTesting
  fun handledForm(form: BookingForm?): Observable<Intent> {

    return Observable.create { subscriber ->
      when {
      // if it's external action, keep loading the web view
        form?.isOAuthForm ?: false || form?.externalAction() != null -> handleArgs(form!!)
        else -> {
          val data = Intent()
          // Form can be null, indicating booking end (auth case).
          data.putExtra(KEY_FORM, form as Parcelable?)
          subscriber.onNext(data)
        }
      }
      subscriber.onComplete()
    }
  }

  private fun handleArgs(form: BookingForm?) {
    this.bookingForm = form
    bookingForm?.let {
      when {
        it.isOAuthForm && it.oAuthLink != null -> url.set(it.oAuthLink.toString())
        it.externalAction() != null -> url.set(it.externalAction())
      }
    }
  }
}